Source/Core/Heap.js

```diff
@@ -0,0 +1,200 @@
+/*global define*/
+define([
+        './Check',
+        './defaultValue',
+        './defined',
+        './defineProperties'
+    ], function(
+        Check,
+        defaultValue,
+        defined,
+        defineProperties) {
+    'use strict';
+
+    /**
+     * Array implementation of a heap.
+     *
+     * @alias Heap
+     * @constructor
+     * @private
+     *
+     * @param {Function} comparator The comparator to use for the heap. If comparator(a, b) is less than 0, sort a to a lower index than b, otherwise sort to a higher index.
+     */
+    function Heap(comparator) {
+        //>>includeStart('debug', pragmas.debug);
+        Check.defined('comparator', comparator);
+        //>>includeEnd('debug');
+
+        this._comparator = comparator;
+        this._array = [];
+        this._length = 0;
+        this._maximumLength = undefined;
+    }
+
+    defineProperties(Heap.prototype, {
+        /**
+         * Gets the length of the heap.
+         *
+         * @memberof Heap.prototype
+         *
+         * @type {Number}
+         * @readonly
+         */
+        length : {
+            get : function() {
+                return this._length;
+            }
+        },
+
+        /**
+         * Gets the internal array.
+         *
+         * @memberof Heap.prototype
+         *
+         * @type {Array}
+         * @readonly
+         */
+        internalArray : {
+            get : function() {
+                return this._array;
+            }
+        },
+
+        /**
+         * Gets and sets the maximum length of the heap.
+         *
+         * @memberof Heap.prototype
+         *
+         * @type {Number}
+         */
+        maximumLength : {
+            get : function() {
+                return this._maximumLength;
+            },
+            set : function(value) {
+                this._maximumLength = value;
+                if (this._length > value && value > 0) {
+                    this._length = value;
+                    this._array.length = value;
+                }
+            }
+        }
+    });
+
+    function swap(array, a, b) {
+        var temp = array[a];
+        array[a] = array[b];
+        array[b] = temp;
+    }
+
+    /**
+     * Resizes the internal array of the heap.
+     *
+     * @param {Number} [length] The length to resize internal array to. Defaults to the current length of the heap.
+     */
+    Heap.prototype.reserve = function(length) {
+        length = defaultValue(length, this._length);
+        this._array.length = length;
+    };
+
+    /**
+     * Update the heap so that index and all descendants satisfy the heap property.
+     *
+     * @param {Number} [index=0] The starting index to heapify from.
+     */
+    Heap.prototype.heapify = function(index) {
+        index = defaultValue(index, 0);
+        var length = this._length;
+        var comparator = this._comparator;
+        var array = this._array;
+        var candidate = -1;
+
+        while (true) {
```

Comment:
Can we change this to a `do...while` loop instead?  I'm not a big fan of `while(true)`.